Task-specific pipeline init args

amyeroberts opened the pull request:
# What does this PR do?

Related to: https://github.com/huggingface/transformers/pull/28439#issue-2074806467

Adds a function to build the pipelines init arguments based on the processing objects it accepts: tokenizer, image processor, feature extractor. 

This replaces `PIPELINE_INIT_ARGS` for the task-specific arguments so as to avoid e.g. `tokenizer` being listed as an input for some models when it's not correct. 

Removes `task` as an input argument for the task-specific pipelines (the task is already specified).


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

amyeroberts committed:
Abstract out pipeline init args

pcuenca commented on src/transformers/pipelines/base.py:
Does it make sense to remove `PIPELINE_INIT_ARGS` to maintain a single version of the string? Or must it be supported for external use?

amyeroberts replied:
I've removed them and added the following on L766 so it's backwards compatible for imports

```py
PIPELINE_INIT_ARGS = build_pipeline_init_args(
    has_tokenizer=True, has_feature_extractor=True, has_image_processor=True, supports_binary_output=True
)
```

pcuenca replied:
Great call!

pcuenca commented on src/transformers/pipelines/mask_generation.py:
```diff
@@ -55,15 +55,21 @@ class MaskGenerationPipeline(ChunkPipeline):
             [`PreTrainedTokenizer`].
-        feature_extractor ([`SequenceFeatureExtractor`]):
-            The feature extractor that will be used by the pipeline to encode the input.
+        image_processor ([`BaseImageProcessor`]):
+            The image processor that will be used by the pipeline to encode data for the model. This object inherits from
+            [`BaseImageProcessor`].
```
Will another `image_processor` snippet be appended with `build_pipeline_init_args`? Same question for `device` below.

amyeroberts replied:
Good point. Yes, they would be. I've removed all the args here and added additional ones to the `add_end_docstring` decorator

pcuenca reviewed (commented):
Very cool, this will be very useful and clarifying!

Narsil commented:
Thanks for this work ! I don't see the PR's documentation though, would be nice to take a look at the rendered stuff.

Also would this be interesting @amyeroberts : https://github.com/huggingface/doc-builder/issues/465 ? (I've seen a discrepancy between docstring and typing in safetensors, and thought this might be interesting in transformers if done in core specific places potentially).

amyeroberts committed:
Address PR comments

amyeroberts committed:
Reword

amyeroberts commented on src/transformers/pipelines/base.py:
```diff
@@ -728,6 +728,62 @@ def predict(self, X):
+    if supports_binary_output:
+        docstring += r"""
+        binary_output (`bool`, *optional*, defaults to `False`):
+            Flag indicating if the output the pipeline should happen in a binary format (i.e., pickle) or as raw text."""
```
Agreed! I've updated to use `serialized format` and `raw output data`

amyeroberts commented on src/transformers/pipelines/base.py:
```diff
@@ -728,6 +728,62 @@ def predict(self, X):
+        docstring += r"""
+        binary_output (`bool`, *optional*, defaults to `False`):
+            Flag indicating if the output the pipeline should happen in a binary format (i.e., pickle) or as raw text."""
+    return docstring
```
Yep! Added now

amyeroberts committed:
BC PIPELINE_INIT_ARGS

amyeroberts committed:
Remove old arguments

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_28439). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

amyeroberts committed:
Small fix

amyeroberts commented:
@pcuenca Thank you for your review! I believe I've addressed all of your comments. 

@Narsil We finally have the docs 🥳 Re typing and docstring checks, I'm definitely pro when the types are given. We currently something which does this in the check_docstrings - [specifically here](https://github.com/huggingface/transformers/blob/708b19eb093bcfb4efcfa229925b249397c2bd5e/utils/check_docstrings.py#L976). However, all of the pipeline classes are ignored because they accept arguments which aren't documented e.g. most accept `tokenizer` even if it isn't used. Once possibility is explicitly setting the documented kwargs in the `__init__` and passing the rest through `**kwargs` in this or a follow up PR. WDYT?

pcuenca commented on src/transformers/pipelines/zero_shot_classification.py:
```diff
@@ -43,7 +43,7 @@ def __call__(self, sequences, labels, hypothesis_template):
 
 
-@add_end_docstrings(build_pipeline_init_args(has_image_processor=True))
+@add_end_docstrings(build_pipeline_init_args(has_tokenizer=True))
```
🙌

pcuenca commented on src/transformers/pipelines/base.py:
```diff
@@ -728,6 +728,62 @@ def predict(self, X):
+    if has_tokenizer:
+        docstring += r"""
+        tokenizer ([`PreTrainedTokenizer`]):
+            The tokenizer that will be used by the pipeline to encode data for the model. This object inherits from
```
Completely unimportant, just pointing out this wasn't incorporated in case you wanted to (based on your previous +1) :)

pcuenca reviewed (approved):
Looks great!

ArthurZucker reviewed (approved):
Thanks for working on this 🤗
